Allow using environment variables instead of configuration.yml

mfrister opened the pull request:
Environment variables are a good idea for automated deployments, e.g. they can't accidentally be checked in and are a operating system concept working everywhere. See here for more reasons: http://12factor.net/config

This also documents the configuration, renames Redmine::Configuration to OpenProject::Configuration and adds a spec.

mfrister committed:
Configuration: make keys overridable via env vars; flatten hierarchy

This results in a new mail configuration format. The old format can
still be used, but shows a deprecation warning.
This also removes support for two even older mail configuration formats.

mfrister committed:
Rename Redmine::Configuration to OpenProject::Configuration

This removes the configuration test, the test will later be updated
and re-added as a spec.

mfrister committed:
Add configuration spec

mfrister committed:
Document configuration and update configuration.yml.example

hschink commented on spec/lib/open_project/configuration_spec.rb:
```diff
@@ -0,0 +1,148 @@
+describe OpenProject::Configuration do
+  # let(:config) { OpenProject::Configuration }
+  describe 'an empty configuration' do
+    let(:config) { load_conf('empty.yml', 'test') }
```
I assume that you're not interested in testing the file loading/parsing capabilities of Rails? Thus, I propose that you either mock the file access or that you separate the file loading from the other aspects.

hschink commented:
I tested the code with `AUTOLOGIN_COOKIE_NAME` and it works! :+1: But which test tests that?

mfrister committed:
Make Configuration spec independent of YAML files, test env vars

mfrister commented:
Restructured the Configuration module a bit to allow testing without the YAML files. Removed them and added a new test for the environment variables.

hschink commented:
Very nice! :cherry_blossom: Final question: Do you need to clean up the `ENV` tests if you use

``` ruby
ENV.stub(:[]).with('SOMESETTING').and_return('bar')
```

?

mfrister commented:
No, but the test also fails:

```
ENV received :[] with unexpected arguments
         expected: ("SOMESETTING")
              got: ("ANSICON")
```

Let's merge :dancer:
